File: Weight_Calories.R

# x(input) - Calorie intake
# y(output) - Weight gain


# Load the dataset

wg.cc <- read.csv(file.choose(), header = T)
colnames(wg.cc) <- c("Weight gain", "Calories")
View(wg.cc)
attach(wg.cc)

# Exploratory data analysis
summary(wg.cc)


install.packages("Hmisc")
library(Hmisc)
describe(wg.cc) # To have a quick glance of a data


install.packages("lattice") # Highly used for data visualization
library("lattice") # dotplot is part of lattice package

# Graphical exploration
dotplot(wg.cc$Calories, main = "Dot Plot of Calories Consumed")
dotplot(wg.cc$`Weight gain`, main = "Dot Plot of Weight Gained")

#Boxplot Representation

boxplot(wg.cc$Calories, col = "dodgerblue4")
boxplot(wg.cc$`Weight gain`, col = "red", horizontal = T)

#Histogram Representation

hist(wg.cc$Calories)
hist(wg.cc$`Weight gain`)

# Normal QQ plot
qqnorm(wg.cc$Calories)
qqline(wg.cc$Calories)

qqnorm(wg.cc$`Weight gain`)
qqline(wg.cc$`Weight gain`)


# Bivariate analysis
# Scatter plot
plot(wg.cc$Calories, wg.cc$`Weight gain`, main = "Scatter Plot", col = "Dodgerblue4", 
     col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Calories Consumed", 
     ylab = "Weight Gained (gms)", pch = 20)  # plot(x,y)


attach(wg.cc)

# Correlation Coefficient
cor(Calories, `Weight gain`)

# Linear Regression model
reg <- lm(`Weight gain` ~ Calories, data = wg.cc) # Y ~ X
summary(reg)

confint(reg, level = 0.95)


pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)

View(pred)


# ggplot for adding Regression line for data
library(ggplot2)

ggplot(data = wg.cc, aes(Calories, `Weight gain`) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ x)


# Evaluation the model for fitness 
cor(pred$fit, wg.cc$`Weight gain`)

reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse


# Transformation Techniques
# transform the variables to check whether the predicted values are better

# input = log(x); output = y

plot(log(Calories), `Weight gain`)
cor(log(Calories), `Weight gain`)

reg_log <- lm(`Weight gain`~ log(Calories), data = wg.cc)
summary(reg_log)

confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")

pred <- as.data.frame(pred)
cor(pred$fit, wg.cc$`Weight gain`)

rmse <- sqrt(mean(reg_log$residuals^2))
rmse

# Regression line for data
ggplot(data = wg.cc, aes(log(Calories), `Weight gain`) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ log(x))



# Log transformation applied on 'y'
# input = x; output = log(y)

plot(Calories, log(`Weight gain`))
cor(Calories, log(`Weight gain`))

reg_log1 <- lm(log(`Weight gain`) ~ Calories, data = wg.cc)
summary(reg_log1)

predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))

pred <- exp(predlog)  # Antilog = Exponential function
pred <- as.data.frame(pred)
cor(pred$fit, wg.cc$`Weight gain`)

res_log1 = `Weight gain` - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse

# Regression line for data
ggplot(data = wg.cc, aes(Calories, log(`Weight gain`)) ) +
  geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)



# Non-linear models = Polynomial models
# input = x & x^2 (2-degree) and output = log(y)

reg2 <- lm(log(`Weight gain`) ~ Calories + I(Calories*Calories), data = wg.cc)
summary(reg2)

predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)

pred <- as.data.frame(pred)
cor(pred$fit, wg.cc$`Weight gain`)

res2 = `Weight gain` - pred$fit
rmse <- sqrt(mean(res2^2))
rmse

# Regression line for data
ggplot(data = wg.cc, aes(Calories, log(`Weight gain`)) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))


# Data Partition

# Random Sampling
n <- nrow(wg.cc)
n1 <- n * 0.8
n2 <- n - n1

train_ind <- sample(1:n, n1)
train <- wg.cc[train_ind, ]
test <-  wg.cc[-train_ind, ]


plot(train$Calories, train$`Weight gain`)
plot(test$Calories, test$`Weight gain`)

model <- lm(`Weight gain` ~ Calories, data = wg.cc) # Y ~ X
summary(model)

confint(reg,level=0.95)

res <- predict(model,interval = "confidence", newdata = test)

predict_original <- as.data.frame(res)
test_error <- test$`Weight gain` - predict_original$fit # calculate error/residual
test_error

test_rmse <- sqrt(mean(test_error^2))
test_rmse

res_train <- predict(reg,interval = "confidence", newdata = train)

predict_original_train <- as.data.frame(res_train)
train_error <- train$`Weight gain` - predict_original_train$fit # calculate error/residual
train_error

train_rmse <- sqrt(mean(train_error^2))
train_rmse

# Alternate Model 

reg <- lm(`Weight gain` ~ Calories+I(Calories*Calories), data = wg.cc)
summary(reg)

confint(reg,level=0.95)
res <- predict(reg,interval = "confidence", newdata = test)

predict_original <- as.data.frame(res)
test_error <- test$`Weight gain` - predict_original$fit # calculate error/residual
test_error

test_rmse <- sqrt(mean(test_error^2))
test_rmse

res_train <- predict(reg,interval = "confidence", newdata = train)

predict_original_train <- as.data.frame(res_train)
train_error <- train$`Weight gain` - predict_original_train$fit # calculate error/residual
train_error

train_rmse <- sqrt(mean(train_error^2))
train_rmse

File: Weight_Calories.py
# Importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

Wgcc = pd.read_csv("C:/Users/personal/Desktop/calories_consumed.csv")
Wgcc.columns = "Wg", "CC" # Wg = Weight Gain & CC = Calories Consumed
Wgcc.columns

# Exploratory data analysis:
Wgcc.describe()

#Graphical Representation
import matplotlib.pyplot as plt # mostly used for visualization purposes 

plt.bar(height = Wgcc.Wg, x = np.arange(1, 15, 1))
plt.hist(Wgcc.Wg) #histogram
plt.boxplot(Wgcc.Wg) #boxplot

plt.bar(height = Wgcc.CC, x = np.arange(1, 15, 1))
plt.hist(Wgcc.CC) #histogram
plt.boxplot(Wgcc.CC) #boxplot

# Scatter plot
plt.scatter(x = Wgcc['CC'], y = Wgcc['Wg'], color = 'red') 

# correlation
np.corrcoef(Wgcc.CC, Wgcc.Wg) 

# Covariance
cov_output = np.cov(Wgcc.CC, Wgcc.Wg)[0, 1]
cov_output


# Import library
import statsmodels.formula.api as smf

# Simple Linear Regression
model = smf.ols('Wg ~ CC', data = Wgcc).fit()
model.summary()

pred1 = model.predict(pd.DataFrame(Wgcc['CC']))

# Regression Line
plt.scatter(Wgcc.CC, Wgcc.Wg)
plt.plot(Wgcc.CC, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res1 = Wgcc.Wg - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1

######### Model building on Transformed Data
# Log Transformation
# x = log(Calories); y = Wg

plt.scatter(x = np.log(Wgcc['CC']), y = Wgcc['Wg'], color = 'brown')
np.corrcoef(np.log(Wgcc.CC), Wgcc.Wg) #correlation

model2 = smf.ols('Wg ~ np.log(CC)', data = Wgcc).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(Wgcc['CC']))

# Regression Line
plt.scatter(np.log(Wgcc.CC), Wgcc.Wg)
plt.plot(np.log(Wgcc.CC), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res2 = Wgcc.Wg - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2


#### Exponential transformation
# x = Calories; y = log(Wg)

plt.scatter(x = Wgcc['CC'], y = np.log(Wgcc['Wg']), color = 'orange')
np.corrcoef(Wgcc.CC, np.log(Wgcc.Wg)) #correlation

model3 = smf.ols('np.log(Wg) ~ CC', data = Wgcc).fit()
model3.summary()

pred3 = model3.predict(pd.DataFrame(Wgcc['CC']))
pred3_at = np.exp(pred3)
pred3_at

# Regression Line
plt.scatter(Wgcc.CC, np.log(Wgcc.Wg))
plt.plot(Wgcc.CC, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res3 = Wgcc.Wg - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3


#### Polynomial transformation
# x = Calories; x^2 = Calories*Calories; y = log(Wg)

model4 = smf.ols('np.log(Wg) ~ CC + I(CC*CC)', data = Wgcc).fit()
model4.summary()

pred4 = model4.predict(pd.DataFrame(Wgcc))
pred4_at = np.exp(pred4)
pred4_at

# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = Wgcc.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)


plt.scatter(Wgcc.CC, np.log(Wgcc.Wg))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()


# Error calculation
res4 = Wgcc.Wg - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4


# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse

###################
# The best model - From the RMSE SLR is the best model with Low Rmse

from sklearn.model_selection import train_test_split

train, test = train_test_split(Wgcc, test_size = 0.2)

finalmodel = smf.ols('Wg ~ CC', data = train).fit()
finalmodel.summary()


# Predict on test data
test_pred = finalmodel.predict(pd.DataFrame(test))
test_pred

# Model Evaluation on Test data
test_res = test.Wg - test_pred
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse


# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
train_pred

# Model Evaluation on train data
train_res = train.Wg -  train_pred
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse


